internal/expr/prepare.go

```diff
@@ -0,0 +1,183 @@
+package expr
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+)
+
+// PreparedExpr represents an SQL expression after the input and output parts
+// have been replaced by their corresponding expansions.
+type PreparedExpr struct {
+	ParsedExpr *ParsedExpr
+	SQL        string
+}
+
+type typeNameToInfo map[string]*info
+
+// prepareInput checks that the input expression corresponds to a Go struct
+// passed to Prepare.
+func prepareInput(ti typeNameToInfo, p *inputPart) error {
+	if inf, ok := ti[p.source.prefix]; ok {
+		if _, ok := inf.tagToField[p.source.name]; ok {
+			return nil
+		}
+		return fmt.Errorf(`there is no tag with name "%s" in "%s"`,
+			p.source.name, inf.structType.Name())
+	}
+	return fmt.Errorf(`unknown type: "%s"`, p.source.prefix)
+}
+
+func starCount(fns []fullName) int {
+	s := 0
+	for _, fn := range fns {
+		if fn.name == "*" {
+			s++
+		}
+	}
+	return s
+}
+
+func prepareOutput(ti typeNameToInfo, p *outputPart) ([]string, error) {
+
+	var outCols []string = make([]string, 0)
+
+	// Check target struct type and its tags are valid.
+	for _, t := range p.target {
+		inf, ok := ti[t.prefix]
+		if !ok {
+			return nil, fmt.Errorf("unknown type: %s", t.prefix)
+		}
+
+		_, ok = inf.tagToField[t.name]
+		if !ok && t.name != "*" {
+			return nil, fmt.Errorf(`there is no tag with name "%s" in "%s"`, t.name, inf.structType.Name())
+		}
+	}
+
+	// Check asterisk are in correct places.
+
+	sct := starCount(p.target)
+	scc := starCount(p.source)
+
+	if sct > 1 || scc > 1 || (scc == 1 && sct == 0) {
+		return nil, fmt.Errorf("invalid asterisk in output expression")
+	}
+
+	starTarget := sct == 1
+	starColumn := scc == 1
+
+	lenS := len(p.source)
+	lenT := len(p.target)
```

Comment:
I think `numSources` and `numTargets` might be clearer?

Reply:
Changed.